substrate-node/pallets/pallet-smart-contract/src/tests.rs

```diff
@@ -1062,6 +1062,100 @@ fn test_node_contract_billing_details() {
+            node_id,
+            generate_deployment_hash(),
+            get_deployment_data(),
+            1,
+            None
+        ));
+        let contract_id = 1;
+
+        push_contract_resources_used(contract_id);
+        push_nru_report_for_contract(contract_id, 10);
+
+        // Ensure contract_id is stored at right billing loop index
+        let index = SmartContractModule::get_billing_loop_index_from_contract_id(contract_id);
+        assert_eq!(
+            SmartContractModule::contract_to_bill_at_block(index),
+            vec![contract_id]
+        );
+
+        let initial_total_issuance = Balances::total_issuance();
+        // advance 24 cycles
+        for i in 1..=24 {
+            let block_number = 1 + i * 10;
+            pool_state.write().should_call_bill_contract(
+                contract_id,
+                Ok(Pays::Yes.into()),
+                block_number,
+            );
+            run_to_block(block_number, Some(&mut pool_state));
+        }
+
+        let twin_balance = Balances::total_balance(&twin.account_id);
+        let total_amount_billed = initial_twin_balance - twin_balance;
+        info!("current balance {:?}", twin_balance);
+
+        info!("total amount billed {:?}", total_amount_billed);
+
+        let staking_pool_account_balance = Balances::free_balance(&get_staking_pool_account());
+        info!(
+            "staking pool account balance, {:?}",
+            staking_pool_account_balance
+        );
+
+        // 5% is sent to the staking pool account
+        assert_eq!(
+            staking_pool_account_balance,
+            (Perbill::from_percent(5) * total_amount_billed)
+        );
+
+        // 10% is sent to the foundation account
+        let pricing_policy = TfgridModule::pricing_policies(1).unwrap();
+        let foundation_account_balance = Balances::free_balance(&pricing_policy.foundation_account);
+        assert_eq!(
+            foundation_account_balance,
+            (Perbill::from_percent(10) * total_amount_billed)
+        );
+
+        // 50% is sent to the sales account
+        let sales_account_balance = Balances::free_balance(&pricing_policy.certified_sales_account);
+        assert_eq!(
+            sales_account_balance,
+            (Perbill::from_percent(50) * total_amount_billed)
+        );
+
+        let total_issuance = Balances::total_issuance();
+        // total issueance is now previous total - amount burned from contract billed (35%)
+        let burned_amount = Perbill::from_percent(35) * total_amount_billed;
+        assert_eq_error_rate!(
+            total_issuance,
+            initial_total_issuance - burned_amount as u64,
+            1
+        );
+
+        // amount unbilled should have been reset after a transfer between contract owner and farmer
+        let contract_billing_info =
+            SmartContractModule::contract_billing_information_by_id(contract_id);
+        assert_eq!(contract_billing_info.amount_unbilled, 0);
```

Comment:
could we create a function here to avoid repeating part of code of previous test ?
maybe something that would take `total_amount_billed` as input

Reply:
I will use the already implemented validate_distribution_rewards() function.